### PR TITLE
Request-level caching for sandboxes

### PR DESCRIPTION
**Pull request by noahmoss**
Adds request-level caching for sandboxes, similar to how we do caching for data permissions.

### CODE REVIEW

**Comment by noahmoss** [comment]
* **#46138** <a href="https://app.graphite.dev/github/pr/metabase/metabase/46138?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @noahmoss and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by escherize — APPROVED** [review]
LGTM

File: enterprise/backend/src/metabase_enterprise/sandbox/api/util.clj [hunk 1]
```diff
@@ -36,11 +36,14 @@
...
-  "Given a user-id and optional set of table-ids, return the sandboxes that should be enforced for the current user on
-  any of the tables. A sandbox is not enforced if the user is in a different permissions group that grants full access
-  to the table."
-  [user-id & [table-ids]]
+(defenterprise enforced-sandboxes-for-user
+  "Given a user-id, returns the set of sandboxes that should be enforced for the provided user ID. This result is cached
+  for the duration of a request in [[metabase.models.data-permissions/*sandboxes-for-user*]].
+
+  WARNING: This should NOT be used directly for sandboxing enforcement. Use `*sandboxes-for-user*` or
+  `enforced-sandboxes-for-tables` below, so that the cache is used."
```
[the remark this answers is not in the document]
Comment (noahmoss) [reply enterprise/backend/src/metabase_enterprise/sandbox/api/util.clj:44]: Hmm — interesting idea but might be overkill for this. I don't think many other consumers (outside our team) will be reaching for these functions. Maybe if this does end up being misused we can look at locking it down more? But I don't hold strong convictions.

[the remark this answers is not in the document]
Comment (escherize) [reply enterprise/backend/src/metabase_enterprise/sandbox/api/util.clj:44]: Yeah, I don't think it's worth it. Just a case where making something harder to call is a good idea (rare). :)

**Pull request merged**